### PR TITLE
Bouncing messages between daemons

### PR DESCRIPTION
**Pull request by cdecker**
The main goal of this is to be able to forward messages between any set of daemons. This includes special purpose daemons, like the gossip daemon, getting the messages from the owner daemons, and getting forwarding messages from special purpose daemons back to the owning daemons.

Since these messages are forwarded without any special handling in between, I created a set of messages that may be sent/received by any daemons, and added handlers for them in the subdaemon and in the main-daemon.

The `forward_gossip_msg` is sent over this path:

     peer_owner -> main_daemon -> gossip_daemon

while the `forward_peer_msg` is currently used to forward gossip announcements like this

    gossip_daemon -> main_daemon -> peer_owner

if the gossip_daemon is not the owner.

### CODE REVIEW

File: lightningd/gossip/gossip.c [hunk 1]
```diff
@@ -198,6 +204,14 @@ static void queue_pkt(struct peer *peer, u8 *msg)
  * queued gossip messages and processes them. */
 static struct io_plan *pkt_out(struct io_conn *conn, struct peer *peer);
 
+/**
+ * peer_dump_gossip_nonlocal - Regularly forward broadcasts to a nonlocal peer
+ * If we are non-local then we just pipe all messages
+ * to the main-daemon synchronously, no need to use
+ * non-blocking IO
```
Comment (rustyrussell) [inline lightningd/gossip/gossip.c]: Or we have a dedicated gossip fd to each peer, and avoid the bouncing altogether?

**Comment by cdecker** [comment]
Thanks @rustyrussell, I think your comments make perfect sense, I quite hated the bouncing through the main-daemon anyway :wink: 

I'll go ahead and rewrite this using direct `gossipd` to `channeld` communication (and later `opening` as well) so that we can forward incoming gossip messages one way and outgoing gossip messages the other way. This has the added advantage of not needing to wrap the forwarded messages (context is given by the receiving `fd`) and we can directly use the peer wire messages.

I think we'll need to hand back a total of 3 fds: `gossipd_in`, `gossipd_out` and `peer_conn_fd`.

As for failing with broken messages I was assuming the other daemons doing a sanity check first (since they have the context to complain to the peer or close the connection) before forwarding, but if we want to protect against broken daemons as well, then we should probably just notify the broken daemon as you suggested.

**Comment by cdecker** [comment]
Minor correction: if I'm not mistaken using the `socketpair` we are actually creating a bidirectional connection anyway, so passing one of the fds from the `socketpair` call should be sufficient to send messages both ways, right?

**Comment by rustyrussell** [comment]
Christian Decker <notifications@github.com> writes:
> Minor correction: if I'm not mistaken using the `socketpair` we are actually creating a bidirectional connection anyway, so passing one of the fds from the `socketpair` call should be sufficient to send messages both ways, right?

Yes, so it's just one for gossipd...

Cheers,
Rusty.

**Comment by cdecker** [comment]
Closing this in favor of #125

**Pull request closed without merging**